nixos/modules/services/hardware/thermald.nix

```diff
@@ -4,27 +4,392 @@ with lib;
+          already present in thermal sysfs, there is no need of a path. Compensation can use
+          min/max and step  size  to  increasing cool the system. Debounce period can be used
+          to force a waiting period for action.
+        '';
+      };
+    };
+  };
+
+  thermalSensorConfig = with types; submodule {
+    options = {
+      Type = mkOption {
+        type = string;
+        description = "The type of this sensor";
+      };
+
+      Path = mkOption {
+        type = nullOr path;
+        default = null;
+        description = "The path to this sensor";
+        example = "/sys/class/thermal/thermal_zone1/";
+      };
+
+      AsyncCapable = mkOption {
+        type = nullOr bool;
+        default = null;
+        description = "if true then we don't need to poll";
+      };
+    };
+  };
+
+  thermalZoneConfig = with types; submodule {
+    options = {
+      Type = mkOption {
+        type = string;
+        description = "The type of thermal zone";
+      };
+
+      Path = mkOption {
+        type = nullOr path;
+        default = null;
+        description = "The path to this thermal zone device";
+      };
+
+      AsyncCapable = mkOption {
+        type = nullOr bool;
+        default = null;
+        description = "if true then we don't need to poll";
+      };
+
+      TripPoints = mkOption {
+        type = (listOf tripPointConfig);
+        description = "The trip points of this thermal zone";
+      };
+    };
+  };
+
+
+  tripPointConfig = with types; submodule {
+    options = {
+      SensorType = mkOption {
+        type = string;
+        description = "The sensor type";
+      };
+
+      Temperature = mkOption {
+        type = int;
+        description = "Temperature at which to take action";
+      };
+
+      Type = mkOption {
+        type = enum [ "max" "passive" "active" ];
+        description = ''
+          If a max type is specified, then daemon will use PID control to
+          aggressively throttle to avoid reaching this temp.
+        '';
+      };
+
+      ControlType = mkOption {
+        type = nullOr (enum [ "SEQUENTIAL" "PARALLEL" ]);
+        default = null;
+        description = ''
+          When a trip point temp is violated, then number of cooling devices can be activated.
+          If control type is SEQUENTIAL then, it will exhaust first cooling device before trying
+          next.
+        '';
+      };
+
+      CoolingDevices = mkOption {
+        type = (listOf tripPointCoolingDeviceConfig);
+        default = [];
+        description = ''
+          The cooling devices to use for this trip point
+        '';
+      };
+    };
+  };
+
+  tripPointCoolingDeviceConfig = with types; submodule {
+    options = {
+      Index = mkOption {
+        type = nullOr int;
+        default = null;
+        description = "The index of this cooling device";
+      };
+
+      Type = mkOption {
+        type = string;
+        description = ''
+          The type of cooling device.
```

Comment:
It's a little bit complicated. From my understanding we have:

- `ThermalZones.Type`: The type of a thermal zone, this is basically the "name" of this zone and I'm not sure if it actually refers to any system capabilities.
- `TripPoint.SensorType`: The sensor used to detect the temperature of this thermal zone. This refers to a sensor under `/sys/class/thermal` so an example could be `x86_pkg_temp`. This can also refer to a sensor defined in the `<ThermalSensors>` section.
- `CoolingDevices.Type`: The identifier of the driver to use to cool this thermal zone. Some of these can be found under `/sys/class/thermal` but some of them seem to be kernel drivers that do not appear under that path. Examples include: `rapl_controller`, `intel_pstate`, `intel_powerclamp`, `cpufreq` and `Processor`